pkg/backend/display/richDiff.go

```diff
@@ -0,0 +1,175 @@
+package display
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/engine"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/plugin"
+	"github.com/pulumi/pulumi/pkg/util/contract"
+)
+
+func parseDiffPath(path string) ([]interface{}, error) {
+	// Complete paths obey the following grammar:
+	//
+	// propertyName := [a-zA-Z_$][a-zA-Z0-9_$]*
+	// arrayIndex := '[' [0-9]+ ']'
+	// propertyAccessor := ( '.' propertyName ) | arrayIndex
+	// path := propertyName { propertyAccessor }
```

Comment:
given that thsi is all done with regexes, i'm actually a little surprised the impl isn't just purely regexes.  That would make it easy to just not have to have this spec+code to keep in sync.  And it would let the regex engine do the matching/capturing for us.

Reply:
> `propertyName { propertyAccessor }`

What do the curlies represent here?

Reply:
These are not regexes, and the recursive nature of the grammar makes things less than clear when regexes are used. This is a very-nearly-EBNF grammar with a bit of regexy stuff thrown in for simplicity's sake; see https://en.wikipedia.org/wiki/Extended_Backus–Naur_form#Table_of_symbols for details. The curlies mean "0 or many".